**Name**: <redacted> <redacted>  
**GitHub URL**: https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>  
**Commit ID**: 6c4a835fd151427bdaf6da365600d81ddea72198

In [179]:
# Imports section
import pandas as pd
from sklearn import model_selection
from sklearn import linear_model
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [180]:
# Using pandas load the dataset (load remotely, not locally)
data = pd.read_csv('science_data_large.csv')
# Output the first 15 rows of the data
display(data.head(15))
# Display a summary of the table information (number of datapoints, etc.)
data.describe()

,Temperature °C,Mols KCL,Size nm^3
0,469,647,6.244743e+05
1,403,694,5.779610e+05
2,302,975,6.196847e+05
3,779,916,1.460449e+06
4,901,18,4.325726e+04
5,545,637,7.124634e+05
6,660,519,7.006960e+05
7,143,869,2.718260e+05
8,89,461,8.919803e+04
9,294,776,4.770210e+05


,Temperature °C,Mols KCL,Size nm^3
count,1000.000000,1000.000000,1.000000e+03
mean,500.500000,471.530000,5.086111e+05
std,288.819436,288.482872,4.474838e+05
min,1.000000,1.000000,1.611429e+01
25%,250.750000,226.750000,1.298267e+05
50%,500.500000,459.500000,3.827182e+05
75%,750.250000,710.250000,7.603211e+05
max,1000.000000,1000.000000,1.972127e+06


## Part 2. Splitting the dataset

In [181]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data.drop('Size nm^3', axis=1)
y = data['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y, test_size=0.1, train_size=0.9, random_state=383)

## Part 3. Perform a Linear Regression

In [182]:
# Use sklearn to train a model on the training set
model = linear_model.LinearRegression().fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
datapoint = [383, 311]
display(f'sample datapoint: {datapoint}')
display(f'prediction based on sample datapoint: {model.predict([datapoint])}')
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
display(f'model score: {model.score(X_test, y_test)}')
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
display(f'model coefficients: {model.coef_}')
display(f'model intercept: {model.intercept_}')

'sample datapoint: [383, 311]'

/Users/<redacted>/Documents/SchoolWork/Year4/cs383/assignment-4-equation-of-a-slime-<redacted>/sklearn-env/lib/python3.9/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


'prediction based on sample datapoint: [237155.12355933]'

'model score: 0.803020654504894'

'model coefficients: [ 895.03563334 1040.66379593]'

'model intercept: -429289.9645449292'

**Report**:  
At a high level and in this context, the score computed by the `score()` method represents how well the data is represented by a linear regression model. Effectively, a score of 1 implies the relationship between the features and label is completely linear, while a score of 0 implies this relationship is not linear at all. Thus, here a score of ~0.80 indicates that the data can somewhat be represented linearly, but not perfectly so.

**Sample equation**: $$s = (895.03563334)t + (1040.66379593)m_{KCl} - 429289.9645449292$$  
where $$s:\text{Size nm$^2$ } \\ t:\text{Temperature \textdegree C } \\ m_{KCl}:\text{Mols KCl }$$

## Part 4. Use Cross Validation

In [183]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
display(f'model scores: {model_selection.cross_val_score(model, X_train, y_train)}')
# Report on their finding and their significance

'model scores: [0.86183095 0.86077975 0.87680024 0.85444433 0.84347539]'

**Report**:  
Here the `cross_val_score()` method is used to generate variations on the dataset to test the model's effectiveness in different scenarios with different (but similar) datapoints. It then scores the model based on these variations. Here we get scores from ~0.84-0.87, which indicate relative consistency in the model's effectiveness, as well as consistency with our initial evaluation above (which was a score of around ~0.80). If we had computed scores way outside of ~0.80 (like 0.10 or 0.30), that indicates that our model is actually not very effective in representing our dataset.

## Part 5. Using Polynomial Regression

In [184]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures()
X_train_poly = poly.fit_transform(X_train, y_train)
poly_model = linear_model.LinearRegression().fit(X_train_poly, y_train)
# Report on the metrics and output the resultant equation as you did in Part 3.
aug_dp = poly.fit_transform([datapoint])
display(f'sample datapoint: {datapoint}')
display(f'augmented datapoint: {aug_dp}')
display(f'prediction based on augmented datapoint: {poly_model.predict(aug_dp)}')
display(f'poly_model score: {poly_model.score(X_train_poly, y_train)}')

display(f'poly_model coefficients: {poly_model.coef_}')
display(f'poly_model intercept: {poly_model.intercept_}')

# below is code I used to generate the equation
feature_names = ['1', 't', 'm_{KCl}', 't^2', '(t)m_{KCl}', 'm_{KCl}^2']
equation = f's = {poly_model.intercept_:.17f}'
for name, coef in zip(feature_names[1:], poly_model.coef_[1:]): # since the first coef is 0, we start at index 1 in both lists
    equation += f' + ({coef:.17f}){name}'

#display(equation)

'sample datapoint: [383, 311]'

'augmented datapoint: [[1.00000e+00 3.83000e+02 3.11000e+02 1.46689e+05 1.19113e+05 9.67210e+04]]'

'prediction based on augmented datapoint: [245585.45713636]'

'poly_model score: 1.0'

'poly_model coefficients: [ 0.00000000e+00  1.20000000e+01 -1.20578939e-07 -2.25769772e-11\n  2.00000000e+00  2.85714287e-02]'

'poly_model intercept: 1.5885336324572563e-05'

**Report**:  
Here the score represents how well the data is represented not by a linear regression model, but the polynomial regression model found using the PolynomialFeatures library. In this case, we get a score of 1, which means the data, or relationship between our features and label, can be perfectly represented by our polynomial regression model (of degree 2, a quadratic function). What this means, in effect, is that the dataset we used to train the model can be perfectly represented/modeled using a quadratic equation, which can be seen below. This is *very* unlikely to happen for any sort of real-world measured dataset, so I am assuming the data for this assignment was generated as a completely quadratic relationship.

**Equation**:  
$$s = 0.00001588533632457 + (12.00000001414722384)t + (-0.00000012057893888)m_{KCl} + \text{ }$$
$$(-0.00000000002257698)t^2 + (2.00000000000424194)(t)m_{KCl} + (0.02857142870073362)m_{KCl}^2$$